### PR TITLE
VCS: Allow to stage and commit all files with Ctrl + Enter

### LINKED ISSUE #2988: Allow to use Ctrl + Enter to stage all files and commit in version control commit dock
### Describe the project you are working on

[Goost](https://goostengine.github.io/) - Godot Engine Extension.

### Describe the problem or limitation you are having in your project

I'm currently working on built-in Git integration in Goost as seen in goostengine/goost#98, and managed to resolve quite a bunch of usability issues so far even with v1.0 of the official Git plugin.

Yet one of the greatest productivity booster for me is having ability to immediately commit stuff. Currently, you have to:
1. Stage all files.
2. Write commit message.
3. Use mouse to press the "Commit" button.

image

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I suggest adding ability to use <kbd>Ctrl + Enter</kbd> shortcut which allows to save the time it takes to commit all files.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Just like in VS Code:

image

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, there's a way to inject special manager object which would listen to input events and delegate that to commit message box, or vice versa. But doing so would mean to rely on implementation detail, which is error-prone.

```c++
void EditorVCSInterfaceGitManager::_input(const Ref<InputEvent> &p_event) {
	ERR_FAIL_NULL(commit_msg_box);

	if (!commit_msg_box->has_focus() || commit_msg_box->get_text().empty()) {
		// Do not allow commits with empty message body either.
		// Unfortunately, `VersionControlEditorPlugin::_send_commit_msg` doesn't validate message text.
		return;
	}
	const Ref<InputEventKey> k = p_event;

	if (k.is_valid() && k->is_pressed()) {
		if (k->get_scancode_with_modifiers() == (KEY_MASK_CMD | KEY_ENTER)) {
			VersionControlEditorPlugin::get_singleton()->call_deferred("_stage_all");
			VersionControlEditorPlugin::get_singleton()->call_deferred("_send_commit_msg");
		}
	}
}

```

Yet I'm not sure whether the same is possible to do via GDScript, might not be possible in fact.

### Is there a reason why this should be core and not an add-on in the asset library?

VCS interface is already part of the editor.

### PR DESCRIPTION
**Pull request by Xrayez**
Closes godotengine/godot-proposals#2988.

See 3.x version of this PR: #50432.

### CODE REVIEW

**Comment by ChronicallySerious** [comment]
Would probably work but I tested this anyway on my Win 10. It would be nice if this shortcut was also configurable i.e. used `ED_SHORTCUT()` but I can't seem to find how it can be used on a `TextEdit` so, for now, this might be the only way to get this functionality in.

**Review by twaritwaikar — APPROVED** [review]
Approving since I tested it and there doesn't seem to be another way to do this. It might be worthwhile to discuss a way to implement shortcuts like these in non-context-menu situations as `ED_SHORTCUT` does. `TextEdit` doesn't seem to have it as of now.

Feel free to disregard this review, I am not too experienced in Editor-foo

**Comment by Xrayez** [comment]
Defined shortcut in editor settings as discussed at Godot contributors chat:

image

Still using signal callback, as in `CodeTextEditor`.

**Comment by Xrayez** [comment]
Reminder: this PR has one approval, would be nice for this to get merged before Godot 3.4 is out, see #50432 for 3.x. Note that Godot beta 2 already includes a couple of enhancements and fixes to VCS I've made in the past weeks.

File: editor/plugins/version_control_editor_plugin.cpp [hunk 1]
```diff
@@ -462,7 +486,9 @@ VersionControlEditorPlugin::VersionControlEditorPlugin() {
 	commit_message->set_custom_minimum_size(Size2(200, 100));
 	commit_message->set_wrap_enabled(true);
 	commit_message->connect("text_changed", callable_mp(this, &VersionControlEditorPlugin::_update_commit_button));
+	commit_message->connect("gui_input", callable_mp(this, &VersionControlEditorPlugin::_commit_message_gui_input));
 	commit_box_vbc->add_child(commit_message);
+	ED_SHORTCUT("version_control/commit", TTR("Commit"), KEY_MASK_CMD | KEY_ENTER);
```
Comment (akien-mga) [inline editor/plugins/version_control_editor_plugin.cpp:491]: Shouldn't this be on `commit_button` (with `set_shortcut()` IIRC)?

Comment (Xrayez) [reply editor/plugins/version_control_editor_plugin.cpp:491]: Note that this directly affects `TextEdit` input. Implementing the same behavior of the "Commit" button would lead to modification of in `_send_commit_msg()`. This will lead to workflow compatibility breakage in 3.x.

Currently, if no files are staged, then the informational hint below the button will display "No files added to stage" message when you press the "Commit" button, preventing the commit. In contrast, this PR at its state allows to commit immediately when typing out commit message, so that pressing <kbd>Ctrl + Enter</kbd> will stage all files automatically for you unless some files were already staged.

In other words, it may end up unexpected for existing users that pressing "Commit" button no longer prevents committing. Some users might still expect and want to stage all files (or some files) themselves manually rather than automatically before committing.

Note that I'm talking primarily in relation to 3.x, where I worked on this initially. The current approach is fairly risk-free, so regressions are unlikely to happen.

**Comment by Xrayez** [comment]
Note that I've initially attempted to make this enhancement in goostengine/goost#98, but it would take me to write three times more code via modules compared to this PR. I'd like to avoid writing the type of code which is prone to crashes in case of Godot's internal API changes in some way.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
